lib/Sema/CSApply.cpp

```diff
@@ -1344,12 +1351,25 @@ namespace {
             resultIsLValue = false;
             base = cs.coerceToRValue(base);
             baseTy = cs.getType(base);
+            // We don't really want to attempt AnyKeyPath application
+            // if we know a more specific key path type is being applied.
+            if (!keyPathTy->isEqual(keyPathExprTy)) {
+              cs.TC.diagnose(base->getLoc(),
+                             diag::expr_smart_keypath_application_type_mismatch,
+                             keyPathExprTy,
+                             baseTy)
+                .highlight(index->getSourceRange());
+            }
```

Comment:
@xedin @rudkx Is there a better alternative to diagnosing during CSApply?

Reply:
@jckarter I'm not 100% on the same page but I think ideally we should produce solutions like that, should it be somehow failing constraint system itself so we even get to CSApply and diagnose it using normal CSDiag process?

Reply:
I agree, but I'm not sure how to make that happen. By the time overload resolution gets to considering the KeyPathApplication as a possibility it seems to have already made up its mind that it can try all superclasses of the concrete key path argument type. Is there a good way to suppress that? I don't think it did that pre-9.0.

Reply:
I'll take a look, maybe there is a way to make it so it discards solution with AnyKeyPath in ranking. Will let you know!

Reply:
Ok, looks like there is a place where we get `AnyKeyPath` from https://github.com/apple/swift/blob/master/lib/Sema/CSSolver.cpp#L743 I think you'd want to put your check in there, WDYT?

Reply:
That looks promising, thanks @xedin! Is it OK if I get this patch in to address the immediate problem and look at that as a follow-up?